runtime/lua/vim/shared.lua

```diff
@@ -98,17 +98,41 @@ end
 --- <pre>
 ---  split(":aa::b:", ":")     --> {'','aa','','b',''}
 ---  split("axaby", "ab?")     --> {'','x','y'}
----  split(x*yz*o, "*", true)  --> {'x','yz','o'}
+---  split("x*yz*o", "*", true)  --> {'x','yz','o'}
+---  split("|x|y|z|", "|", true, true) --> {'x', 'y', 'z'}
 --- </pre>
---
---@see |vim.gsplit()|
 ---
---@param s String to split
---@param sep Separator string or pattern
---@param plain If `true` use `sep` literally (passed to String.find)
---@returns List-like table of the split components.
-function vim.split(s,sep,plain)
-  local t={} for c in vim.gsplit(s, sep, plain) do table.insert(t,c) end
+---@see |vim.gsplit()|
+---
+---@param s String to split
+---@param sep Separator string or pattern
+---@param plain If `true` use `sep` literally (passed to String.find)
+---@param trimempty If `true` remove empty items from the front and back of the list
+---@returns List-like table of the split components.
+function vim.split(s, sep, plain, trimempty)
```

Comment:
vimscript split() calls the param `keepempty`. Should we stick with that? Or is there precedent for "trimempty" elsewhere?

Reply:
I mentioned this in the original PR comment, but the discontinuity here is to preserve backward compatibility. If we called it `keepempty` and changed the default behavior we would risk breaking a lot of existing usage of `vim.split`.

Reply:
sorry I missed that. Since there are conflicts anyway, let's rename it to `noempty`. "Trim" potentially useful for a feature where each item is trimmed (e.g. [.NET has this](https://docs.microsoft.com/en-us/dotnet/api/system.stringsplitoptions?view=net-5.0#fields).

.NET also uses the negative case (RemoveEmptyEntries) so there's that :)

Reply:
It’s not “noempty” though, only empty elements from the front and back of the array are removed. “noempty” implies *all* empty elements are removed.

Reply:
oh my. ok, fair enough. Looks like the tests cover that case?

We can go ahead with this PR, though it would be a good time to introduce a "kwargs" variant instead of adding a 4th arg. But we can wait on that, up to you.

Reply:
I think a kwargs variant is a good idea. If we go that route it would make sense to also make `{plain}` a keyword arg so that we don't mix positional and keyword arguments, but we can't do this without breaking backward compatibility. We could make `{plain}` a keyword arg as well as keeping it as a positional one for now to avoid breaking existing scripts, but update the documentation to indicate that new usages should use the keyword variant.

but maybe I'm overthinking this :) what do you think?

Reply:
> kwargs variant is a good idea. If we go that route it would make sense to also make `{plain}` a keyword arg so that we don't mix positional and keyword arguments, but we can't do this without breaking backward compatibility

typicall pattern is to accept exactly 1 arg for the kwargs variant. That's backward-compatible because normally arg1 is a string, not a table. See e.g. screen:expect() https://github.com/neovim/neovim/blob/b854e3ee8c7234f0e144c5e850c3cea9c3e01388/test/functional/ui/screen.lua#L315-L319

It's a bit noisy in the impl, but we can start to build common util functions since this is a pattern that should be pervasive (or just start kwargs from the beginning for most functions).